src/Settings/Builders/RunTest.hs

```diff
@@ -0,0 +1,113 @@
+module Settings.Builders.RunTest (runTestBuilderArgs) where
+
+import Hadrian.Utilities
+import Hadrian.Haskell.Cabal
+
+import Flavour
+import Rules.Test
+import Settings.Builders.Common
+import Settings.Builders.Ghc
+
+-- Arguments to send to the runtest.py script.
+runTestBuilderArgs :: Args
+runTestBuilderArgs = mconcat
+    [ builder RunTest ? do
+        pkgs     <- expr $ stagePackages Stage1
+        libTests <- expr $ filterM doesDirectoryExist $ concat
+                [ [ pkgPath pkg -/- "tests", pkgPath pkg -/- "tests-ghc" ]
+                | pkg <- pkgs, isLibrary pkg, pkg /= rts, pkg /= libffi ]
+
+        debugged          <- ghcDebugged <$> expr flavour
+
+        withNativeCodeGen <- expr ghcWithNativeCodeGen
+        withInterpreter   <- expr ghcWithInterpreter
+        unregisterised    <- expr $ flag GhcUnregisterised
+
+        withSMP           <- expr ghcWithSMP
+
+        windows  <- expr windowsHost
+        darwin   <- expr osxHost
+
+        threads  <- shakeThreads <$> expr getShakeOptions
+        verbose  <- shakeVerbosity <$> expr getShakeOptions
+
+        top      <- expr topDirectory
+        compiler <- expr $ builderPath $ Ghc CompileHs Stage2
+        ghcPkg   <- expr $ builderPath $ GhcPkg Update Stage1
+        haddock  <- expr $ builderPath $ Haddock BuildPackage
+        hp2ps    <- expr $ builderPath $ Hp2Ps
+        hpc      <- expr $ builderPath $ Hpc
+
+        let ifMinGhcVer ver opt = do v <- expr ghcCanonVersion
+                                     if ver <= v then pure opt
+                                                 else pure ""
+
+        -- TODO read environment's ghc flags.
+        --
+        -- See: https://github.com/ghc/ghc/blob/master/testsuite/mk/test.mk#L28
+        let ghcExtraFlags = if unregisterised
+                               then " -optc-fno-builtin"
+                               else ""
+
+        -- Take flags to send to the Haskell compiler from test.mk.
+        -- See: https://github.com/ghc/ghc/blob/master/testsuite/mk/test.mk#L37
+        ghcFlags <- unwords <$> sequence
+            [ pure " -dcore-lint -dcmm-lint -no-user-package-db -rtsopts"
+            , pure ghcExtraFlags
+            , ifMinGhcVer "711" "-fno-warn-missed-specialisations"
+            , ifMinGhcVer "711" "-fshow-warning-groups"
+            , ifMinGhcVer "801" "-fdiagnostics-color=never"
+            , ifMinGhcVer "801" "-fno-diagnostics-show-caret"
+            , pure "-dno-debug-output"
+            ]
+
+        timeout_prog <- (-/- "test" -/- "bin" -/- "timeout") <$> expr buildRoot
```

Comment:
This would be more typical: `expr buildRoot <&> "test/bin/timeout"`.

Reply:
Fixed.